### ISSUE TITLE
mat: VecDense.CopyVec is not safe for overlapping vectors

### ISSUE THREAD

**Issue by kortschak** [gonum/gonum#324]
The underlying `Dcopy` allows direction to be specified by changing the sign of the increments, but we don't expose this to the users of mat. Nor do we check for shadowing, but we do say that it works similarly to the `copy` built-in, which might lead people to believe that [this kind of code](https://play.golang.org/p/spcYwRiRdV) is safe:

```
package main

import (
	"fmt"
)

func main() {
	backing := []int{1, 2, 3, 4, 5, 6, 7, 8, 9}
	a := backing[1:9]
	b := backing[0:8]
	copy(a, b)
	fmt.Println(backing)
}
```

It's not:
```
package main

import (
	"fmt"

	"gonum.org/v1/gonum/mat"
)

func main() {
	// Get vectors with an inc of 2 so we aren't using the copy built-in.
	backing := []float64{1, 0, 2, 0, 3, 0, 4, 0, 5, 0, 6, 0, 7, 0, 8, 0, 9, 0}
	m := mat.NewDense(9, 2, backing)
	a, b := &mat.VecDense{}, &mat.VecDense{}
	a.ColViewOf(m, 0)
	a = a.SliceVec(1, 9)
	b.ColViewOf(m, 0)
	b = b.SliceVec(0, 8)
	a.CopyVec(b)
	fmt.Println(backing)
}
```

```
$ go run badveccopy.go 
[1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0]
```

A safe copy depends on knowing the relative arrangement of elements, we can know this in mat, though that code is not in blas64 which is the last place where we have control, and probably where this should be dealt with since it is the package making the assumption of element ordering.

**Comment by kortschak** [comment]
This is also true for (*Dense).Copy.

@btracey Do you have a view on this?

**Comment by kortschak** [comment]
*ping*

**Comment by btracey** [comment]
If the complexity isn't very high, I think it's best if we can make it behave like the built-in copy. I think we could do that by checking shadowing and then if it's shadowed taking a different path?

**Comment by kortschak** [comment]
For vectors it should be fairly straightforward; three steps, check for overlap, no overlap -> fast path, defined overlap (matched increment) -> define copy direction correctly and otherwise find single element collision (this may happen if increments don't match, but there is overlap) -> define copy direction.

The first two parts are easy, I need to think about an efficient way of determining the last one.

Dense matrices are dimensionally more complicated, so I won't think about them until I have a workable solution for vectors since that may suggest approaches.

**Comment by kortschak** [comment]
OK, for the vector case, you can find if there are any solutions to the diophantine equation s_a i + s_b j = a_0 + b_0 where s_x is the inc of x, x_0 is the address of the first element of x, and 0 <= i < a_n and 0 <= j < b_n. We can find these by finding if GCD(s_a, s_b)|(a_0 + b_0) exists. If it does not there is no overlapping element. If it does, we need to check whether the i and j constraints are met - if not, there is no overlap. If we pass these two tests, we then need to determine the correct direction; this is done by choosing based on which if i or j is lower for the first or last solutions within their ranges.

I think the matrix equivalent will be significantly harder to do even as efficiently as this (noting that this is not wildly efficient).

**Comment by kortschak** [comment]
The solution above depends on allocations to find the initial Diophantine solutions, which is not happy-making.

Of course, the situation becomes significantly simpler if we do not try to handle differing strides/increments. If we change the docs to the effect that it works like the built-in copy when strides are the same for matrices and increments are the same for vectors, and otherwise panics. Then we solve the issue easily. What do you think?

**Comment by btracey** [comment]
I think that's reasonable. In general we panic for overlap, and so in general this will match that behavior. This also allows the special cases of copying one sub-matrix into another sub-matrix, or copying elements of the same vector within itself (because the strides are the same). What it doesn't allow is people being really tricky with their data (from what I see, casting the same backing array to different sizes), which seems fine to me.

**Comment by kortschak** [comment]
Yes, that was my view. I briefly considered having unit increment versus any increment being a special case on the basis that it might be useful to copy a row to a column and vice versa, but I don't think the added complexity is worth it (and usually this would be a swap operation anyway).
